pkg/build/webhook/webhook.go

```diff
@@ -47,53 +52,86 @@ func GitRefMatches(eventRef, configRef string, buildSource *buildapi.BuildSource
-	if len(buildTriggers) == 0 {
-		err = ErrHookNotEnabled
-	}
-	return buildTriggers, err
+// NewWarning returns an StatusError object with a http.StatusOK (200) code.
+func NewWarning(message string) *kerrors.StatusError {
+	return &kerrors.StatusError{ErrStatus: metav1.Status{
+		Status:  metav1.StatusSuccess,
+		Code:    http.StatusOK,
+		Message: message,
+	}}
 }
 
-// ValidateWebHookSecret validates the provided secret against all currently
-// defined webhook secrets and if it is valid, returns its information.
-func ValidateWebHookSecret(webHookTriggers []buildapi.BuildTriggerPolicy, secret string) (*buildapi.WebHookTrigger, error) {
-	for _, trigger := range webHookTriggers {
-		switch trigger.Type {
-		case buildapi.GenericWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.GenericWebHook.Secret), []byte(secret)) {
-				continue
-			}
-			return trigger.GenericWebHook, nil
-		case buildapi.GitHubWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.GitHubWebHook.Secret), []byte(secret)) {
-				continue
+// CheckSecret tests the user provided secret against the secrets for the webhook triggers, if a match is found
+// then the corresponding webhook trigger is returned.
+func CheckSecret(namespace, userSecret string, triggers []*buildapi.WebHookTrigger, secretsClient kcoreclient.SecretsGetter) (*buildapi.WebHookTrigger, error) {
+	for i := range triggers {
+		secretRef := triggers[i].SecretReference
+		secret := triggers[i].Secret
+		if len(secret) > 0 {
+			if hmac.Equal([]byte(secret), []byte(userSecret)) {
+				return triggers[i], nil
 			}
-			return trigger.GitHubWebHook, nil
-
-		case buildapi.GitLabWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.GitLabWebHook.Secret), []byte(secret)) {
-				continue
+		}
+		if secretRef != nil {
+			glog.V(4).Infof("Checking user secret against secret ref %s", secretRef.Name)
+			s, err := secretsClient.Secrets(namespace).Get(secretRef.Name, metav1.GetOptions{})
+			if err != nil && !kerrors.IsNotFound(err) {
+				return nil, err
 			}
-			return trigger.GitLabWebHook, nil
-
-		case buildapi.BitbucketWebHookBuildTriggerType:
-			if !hmac.Equal([]byte(trigger.BitbucketWebHook.Secret), []byte(secret)) {
-				continue
+			if v, ok := s.Data[buildapi.WebHookSecretKey]; ok {
+				if hmac.Equal(v, []byte(userSecret)) {
+					return triggers[i], nil
+				}
 			}
-			return trigger.BitbucketWebHook, nil
 		}
 	}
+	glog.V(4).Infof("did not find a matching secret")
 	return nil, ErrSecretMismatch
 }
 
-// NewWarning returns an StatusError object with a http.StatusOK (200) code.
-func NewWarning(message string) *kerrors.StatusError {
-	return &kerrors.StatusError{ErrStatus: metav1.Status{
-		Status:  metav1.StatusSuccess,
-		Code:    http.StatusOK,
-		Message: message,
-	}}
+func GenerateBuildTriggerInfo(revision *buildapi.SourceRevision, hookType, secret string) (buildTriggerCauses []buildapi.BuildTriggerCause) {
+	hiddenSecret := fmt.Sprintf("%s***", secret[:(len(secret)/2)])
```

Comment:
I know this is not new, but I wonder why we do this?  It doesn't seem at all wise.

Reply:
what part of it doesn't seem wise?

Reply:
(that we're displaying any part of the user provided secret?)

Reply:
allowing half of a secret to end up in plaintext in the BuildTriggerCause doesn't seem brilliant to me.